File: Difficulty: Medium/K Sized Subarray Maximum/k-sized-subarray-maximum.java
//{ Driver Code Starts
//Initial template for JAVA

import java.util.*;
import java.io.*;
import java.lang.*;

class Main
{
    public static void main(String args[])
    {
        //taking input using class Scanner
        Scanner sc = new Scanner(System.in);
        
        //taking total count of testcases
        int t = sc.nextInt();
        
        
        
        while(t-- > 0)
        {
            //taking total number of elements
            int n = sc.nextInt();
            
            //taking size of subArrays 
            int k = sc.nextInt();
            
            //Declaring and Intializing an array of size n
            int arr[] = new int[n];
            
            //adding all the elements to the array 
            for(int i = 0; i <n; i++)
            {
                arr[i] =sc.nextInt();
            }
            
            //Calling the method max_of_subarrays of class solve
            //and storing the result in an ArrayList
            ArrayList <Integer> res = new Solution().max_of_subarrays(arr, n, k);
            
            //printing the elements of the ArrayList
            for (int i = 0; i < res.size(); i++)
                System.out.print (res.get (i) + " ");
            System.out.println();
        }
    }
}
// } Driver Code Ends


//User function template for JAVA
class Solution
{
    //Function to find maximum of each subarray of size k.
    static ArrayList <Integer> max_of_subarrays(int arr[], int n, int k)
    {
        ArrayList <Integer> list= new ArrayList<>(); 
        
        // for(int i=0; i<=n-k; i++){
        //     int max= arr[i]; 
        //     for(int j=i; j<k+i; j++){
        //         if(arr[j]> max){
        //             max=arr[j]; 
        //         }
        //     }
        //     list.add(max); 
        // }
        
        // return list;
        
        
        int i=0; // startign idx of window
        int j=k-1; //  ending index of window
        int prevMaxIdx=-1; // index of Previous max elem
        int max= Integer.MIN_VALUE; 
        
        for(int idx=i; idx<=j; idx++){
            if(arr[idx]>max){
                max=arr[idx]; 
                prevMaxIdx= idx; 
            }
        }
        
        list.add(max); 
        
        while(j<n-1){
            
            // sliding the window
            i++; 
            j++; 
            
            // now checking previous maximum is still in this window or not
            if(prevMaxIdx>=i){
                if(arr[j]<max){
                    list.add(max); 
                    continue; 
                }
                
                // now again finding the max elem in the current window
                else{
                    max=Integer.MIN_VALUE; 
                    for(int idx=i; idx<=j; idx++){
                        if(arr[idx]>max){
                            max=arr[idx];
                            prevMaxIdx = idx; 
                        }
                    }
                    list.add(max); 
                }
            }else{
                max= Integer.MIN_VALUE; 
                for(int idx= i; idx<=j; idx++){
                    if(arr[idx]>max){
                        max=arr[idx]; 
                        prevMaxIdx= idx; 
                    }
                }
                list.add(max); 
            }
        }
        return list;
    }
}

File: Difficulty: Medium/K Sized Subarray Maximum/k-sized-subarray-maximum.cpp
//{ Driver Code Starts
// Initial Template for C++
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
// User function template for C++

// User function template for C++

class Solution {
  public:
    // Function to find maximum of each subarray of size k.
    vector<int> max_of_subarrays(int k, vector<int> &arr) {
        // your code here
        int n=arr.size();
        vector<int> ans;
        // int maxi=arr[0];
        // int prev2=arr[1];
        // int prev1=arr[2];
        
        // maxi=max(maxi,max(prev1,prev2));
        
        // ans[0]=maxi;
    
        // int j=1;
        
        // for(int i=0;i<n;i++){
        //     maxi=max(arr[i],max(prev1,prev2));
        //         ans[j]=maxi;
        //         j++;
        //         prev1=arr[i];
        //         prev2=arr[i-1];
            
        // }
        deque<int>dq;
        for(int i=0;i<n;i++){
            
            //remove elements from front of deque
            if(!dq.empty() && dq.front()==i-k){
                dq.pop_front();
            }
            
            while(!dq.empty() && arr[dq.back()]<=arr[i]){
                dq.pop_back();
            }
            
            dq.push_back(i);
            
            if(i>=k-1){
                ans.push_back(arr[dq.front()]);
            }
        }
        
        return ans;
    }
};

//{ Driver Code Starts.

int main() {

    int t;
    cin >> t;
    cin.ignore();

    while (t--) {

        string ks;
        getline(cin, ks);
        int k = stoi(ks);
        vector<int> arr;
        string input;
        getline(cin, input);
        stringstream ss(input);
        int number;
        while (ss >> number) {
            arr.push_back(number);
        }
        Solution obj;
        vector<int> res = obj.max_of_subarrays(k, arr);
        for (int i = 0; i < res.size(); i++)
            cout << res[i] << " ";
        cout << endl;
    }

    return 0;
}

// } Driver Code Ends